Add Repeat and Retry functionality to sigh test

thebrianchen opened the pull request:
Adding the option to allow retries and repeat tests. I initially tried adding repeat functionality directly into mocha by encapsulating `runner.run()` which would allow test pass/fail/end hooks to be added more easily. However, I ran into a mocha implementation [issue](https://github.com/mochajs/mocha/issues/2706) that requires the entire mocha instance to be re-instantiated and the 'require' cache cleared in between runs. 

I removed the return from `test()` -- not sure where it's being used.

Also, not sure if the package.lock should be excluded // how to auto-exclude it from future commits.

thebrianchen commented:
I signed it!

lindner commented on tools/sigh.js:
```diff
@@ -581,20 +589,23 @@ function test(args) {
-        runner
-      ],
-      {stdio: 'inherit'});
+  for (var i = 0; i < JSON.stringify(options.repeat || 1); i++) {
```
Needs some simple docs.  A comment will do.

Can you hoist the options.repeat value into a constant.  That will make it obvious what's happening,

lindner commented on tools/sigh.js:
```diff
@@ -581,20 +589,23 @@ function test(args) {
-      ],
-      {stdio: 'inherit'});
+  for (var i = 0; i < JSON.stringify(options.repeat || 1); i++) {
+    console.log('RUN %s:', i+1);
```
Optional: consider outputting a timestamp.

lindner commented on devtools/package-lock.json:
```diff
@@ -52,11 +52,6 @@
         "@polymer/polymer": "^3.0.0"
```
omit this file from your PR.  Thanks!

lindner commented:
Check the return-code you're generating?  If any test fails you should return non-zero for the test phase. 

```
😱
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! arcs@0.0.0 test-sigh: `cross-env tools/sigh`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the arcs@0.0.0 test-sigh script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
npm ERR! A complete log of this run can be found in:
npm ERR!     /home/travis/.npm/_logs/2019-01-15T21_18_55_679Z-debug.log
```

lindner commented on tools/sigh.js:
```diff
@@ -536,6 +538,12 @@ function test(args) {
         mocha.suite.emit('require', null, '${test}', mocha);
         mocha.suite.emit('post-require', global, '${test}', mocha);
       `);
+      if (options.retries) {
```
validate value?

also minimist docs say:

""Numeric-looking arguments will be returned as numbers unless opts.string or opts.boolean is set for that argument name."""

thebrianchen replied:
I added validation just now to ensure that is a number. I tried adding a `console.log()` for non-numeric inputs, but logging it in `buildTestRunner()` would print it multiple times.

lindner commented:
LGTM, Raul for approval.

raulverag commented:
I don't see why we would want to repeat or retry tests within one run. What does it even mean if the tests pass on one repeat and then fail on another, other than that the tests are flaky? Furthermore, it seems that if the last run succeeds, the user won't even see any failing output, because that will have scrolled off the screen. Correct me if I'm mistaken, and please elaborate as to what purpose repeats and retries serve.

shans commented:
Friendly reminder to please discuss changes (such as this) which have the potential to change development practices for all of us in a team meeting, ideally before starting development.

lindner commented:
For folks tuning in, this work was done to specifically address issue  #779

My apologies for not bringing this up in a larger context.  It was deemed something simple enough as a good-first-issue.

lindner commented:
Raul suggested that this needs to have summary output at the end.

raulverag commented:
Yes. If the idea is to help diagnose flaky tests, as described in #779, then the results must be displayed in some kind of summary, e.g. "x out of y tests succeeded. First failure was repeat N.", or in the case of retries, "Tests succeeded after X retries".

thebrianchen commented:
The mocha.js implementation of `retries` does not allow for logging and will simply run the test until it either passes or the retry count is exhausted, so I removed the the `retries` flag option.

Added a summary to the end of the runs. Example output:
```
100 runs completed. 4 runs failed.
Failed runs: [4, 47, 49, 81]
```

lindner commented:
LGTM, Raul for final approvals.
